### ISSUE TITLE
Add `OpenAI` JSON Mode

### ISSUE THREAD

**Issue by rlouf** [dottxt-ai/outlines#1305]

**Comment by lhoestq** [comment]
Hi remi ! do you mean `response_format={"type": "json_object"}` ?

I was recently making a demo using [lotus-ai] https://github.com/guestrin-lab/lotus to process pandas DataFrames with a model on HF but it seems to require JSON mode in `outlines`.

I also noticed https://github.com/dottxt-ai/outlines/issues/707 which seems to imply that it could be possible to have a workaround for this feature but it's missing documentation

**Comment by rlouf** [comment]
Yes that's what I mean. We have experimental support for JSON mode in outlines. It's a bit hidden in the documentation: https://dottxt-ai.github.io/outlines/latest/reference/generation/cfg/

**Comment by RobinPicard** [comment]
What do you think about creating a custom type `JsonMode` that would:
- be used to return the appropriate option parameters for the model thanks to the `dispatch` of the `TypeAdapter` for API models
- be turned into a `CFG` during the `Generator` instantiation for local models

**Comment by RobinPicard** [comment]
Ah yes, that's a better idea to use a native type instead of inventing a new one!
